### PR TITLE
Create main.py, cross_validation for multisite_neuroimg_adapt example

### PR DESCRIPTION
**Pull request by alanbijuthomas**
### Description
Create main.py to multisite_neuroimg_adapt example and cross_validation module for reusing leave-one-group-out function.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] In-line docstrings updated.
- [x] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

**Review by shuo-zhou — APPROVED** [review]
Look good to me

File: examples/multisite_neuroimg_adapt/main.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+from sklearn.linear_model import RidgeClassifier
+from sklearn.metrics import accuracy_score
+from sklearn.preprocessing import OneHotEncoder
+
+from kale.interpret import visualize
+from kale.pipeline.multi_domain_adapter import CoIRLS
+from kale.utils.download import download_file_by_url
+
+
+def cross_validation(x, y, covariates, estimator, domain_adaptation=False):
```
Comment (haipinglu) [inline examples/multisite_neuroimg_adapt/main.py:30]: @shuo-zhou Sure this is not reusable?  To discuss this before merging.

**Review by haipinglu — CHANGES_REQUESTED** [review]
To discuss the cv part with @shuo-zhou

File: examples/multisite_neuroimg_adapt/main.py (hunk 1 above)
Comment (shuo-zhou) [reply examples/multisite_neuroimg_adapt/main.py:30]: Other pipelines with dataloader and lightning trainer do not need this

Comment (haipinglu) [reply examples/multisite_neuroimg_adapt/main.py:30]: The issue is not whether some other APIs do not need this. It is about what some APIs/examples may need. A library's value is about what it can do, not what it cannot do.

File: examples/multisite_neuroimg_adapt/main.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        results["Target"].append(tgt)
+        results["Num_samples"].append(x_tgt.shape[0])
+
+    mean_acc = sum([results["Num_samples"][i] * results["Accuracy"][i] for i in range(n_covariates)])
+    mean_acc /= x.shape[0]
+    results["Target"].append("Average")
+    results["Num_samples"].append(x.shape[0])
+    results["Accuracy"].append(mean_acc)
+
+    return pd.DataFrame(results)
```
Comment (haipinglu) [inline examples/multisite_neuroimg_adapt/main.py:62]: As discussed, this should go into `kale.evaluate`

Comment (shuo-zhou) [reply examples/multisite_neuroimg_adapt/main.py:62]: A module `cross_validation` can be created under `kale.evaluate`. This function can be called `leave_one_group_out_cross_validate`.

The source code of these two APIs in scikit-learn can be helpful:

- [LeaveOneGroupOut](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.LeaveOneGroupOut.html#sklearn.model_selection.LeaveOneGroupOut)
- [cross_validate](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html#sklearn.model_selection.cross_validate)

File: kale/evaluate/cross_validation.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            - 'Target': List of unique target groups or classes.
+            - 'Num_samples': List containing number of samples in each target group.
+            - 'Accuracy': List of accuracy scores for each target group.
+    """
+    target, num_samples, accuracy = [], [], []
+    enc = OneHotEncoder(handle_unknown="ignore")
+    covariate_mat = enc.fit_transform(covariates.reshape(-1, 1)).toarray()
+    unique_covariates = np.unique(covariates)
+
+    for tgt in unique_covariates:
```
Comment (shuo-zhou) [inline kale/evaluate/cross_validation.py:29]: Try if the [LeaveOneGroupOut API](https://scikit-learn.org/stable/modules/cross_validation.html#leave-one-group-out) in scikit-learn can be leveraged here.

File: kale/evaluate/cross_validation.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    Args:
+        x: Input data [n_samples, n_features].
+        y: Target labels [n_samples].
+        covariates: Covariates indicating classes for LOOCV [n_samples].
+        estimator: Machine learning estimator to be evaluated from kale or scikit-learn.
+        domain_adaptation: Whether to use domain adaptation during training.
+
+    Returns:
+        dict: A dictionary containing LOOCV results for each target group with 3 keys.
```
Comment (shuo-zhou) [inline kale/evaluate/cross_validation.py:18]: It is leave one group out but not leave one out cross validation here

**Review by shuo-zhou — CHANGES_REQUESTED** [review]

**Review by shuo-zhou — APPROVED** [review]
Look good to me

File: examples/multisite_neuroimg_adapt/main.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+from kale.evaluate import cross_validation
+from kale.pipeline.multi_domain_adapter import CoIRLS
+
+
+def main():
+    # ---- Ignore warnings ----
+    warnings.filterwarnings("ignore")
+
+    # ---- Path to `.yaml` config file ----
+    cfg_path = "configs/tutorial.yaml"
```
Comment (haipinglu) [inline examples/multisite_neuroimg_adapt/main.py:32]: Take a look at the main and arg_parse() of some recent examples (e.g. avmnist, multiomics, polypharmacy) and do it in a similar way. Typically, we do not want to hardcode the config file.

File: examples/multisite_neuroimg_adapt/main.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+from nilearn.datasets import fetch_abide_pcp
+from sklearn.linear_model import RidgeClassifier
+
+from kale.evaluate import cross_validation
+from kale.pipeline.multi_domain_adapter import CoIRLS
+
+
+def main():
+    # ---- Ignore warnings ----
+    warnings.filterwarnings("ignore")
```
Comment (haipinglu) [inline examples/multisite_neuroimg_adapt/main.py:29]: Why this? Many other main.py files do no have it so why must here?

File: examples/multisite_neuroimg_adapt/main.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+    # ---- Path to `.yaml` config file ----
+    cfg_path = "configs/tutorial.yaml"
+    cfg = get_cfg_defaults()
+    cfg.merge_from_file(cfg_path)
+    cfg.freeze()
+    print(cfg)
```
Comment (haipinglu) [inline examples/multisite_neuroimg_adapt/main.py:36]: No seed? Sure, no randomness?

File: examples/multisite_neuroimg_adapt/main.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    # ---- Fetch ABIDE fMRI timeseries ----
+    root_dir = cfg.DATASET.ROOT
```
Comment (haipinglu) [inline examples/multisite_neuroimg_adapt/main.py:39]: Consider whether these new variables are necessary. Take a look at other examples. You did not use it at all places either.

File: kale/evaluate/cross_validation.py
Comment (haipinglu) [inline kale/evaluate/cross_validation.py:1]: Always good to have some docs at the top of a module, like readme.

File: kale/evaluate/cross_validation.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+            - 'Target': List of unique target groups or classes.
+            - 'Num_samples': List containing number of samples in each target group.
+            - 'Accuracy': List of accuracy scores for each target group.
+    """
+    enc = OneHotEncoder(handle_unknown="ignore")
+    group_mat = enc.fit_transform(groups.reshape(-1, 1)).toarray()
+    target, num_samples, accuracy = np.unique(groups).tolist(), [], []
+
+    logo = LeaveOneGroupOut()
```
Comment (haipinglu) [inline kale/evaluate/cross_validation.py:29]: This variable name is problematic since most will read it as logo.

**Review by haipinglu — CHANGES_REQUESTED** [review]
It's almost there but needs some tweaking. Please discuss this with Xianyuan and Shuo for the next iteration. Thanks.

@xianyuanliu Please either do your version of a review for this PR or do a review for the next round before I do my review. Please take a look at the review comments that I made, as this one is quite manageable. I will need you to do more PyKale reviews in future. Many thanks.

File: examples/multisite_neuroimg_adapt/main.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
+"""
+Autism Detection: Domain Adaptation for Multi-Site Neuroimaging Data Analysis
+
+Reference:
```
Comment (xianyuanliu) [inline examples/multisite_neuroimg_adapt/main.py:4]: please keep all references in a consistent format, especially capitalization and author name.

File: examples/multisite_neuroimg_adapt/main.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+    # ---- Fetch ABIDE fMRI timeseries ----
+    fetch_abide_pcp(
+        data_dir=cfg.DATASET.ROOT,
+        pipeline=cfg.DATASET.PIPELINE,
+        band_pass_filtering=True,
+        global_signal_regression=False,
+        derivatives=cfg.DATASET.ATLAS,
+        quality_checked=False,
+        SITE_ID=cfg.DATASET.SITE_IDS,
```
Comment (xianyuanliu) [inline examples/multisite_neuroimg_adapt/main.py:54]: I haven't found the parameter `SITE_ID` in the function `fetch_abide_pcp`. Please let me know if I miss it somehow.

File: kale/evaluate/cross_validation.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    Args:
+        x: Input data [n_samples, n_features].
+        y: Target labels [n_samples].
+        groups: Group labels to be left out [n_samples].
+        estimator: Machine learning estimator to be evaluated from kale or scikit-learn.
+        domain_adaptation: Whether to use domain adaptation during training.
+
+    Returns:
+        dict: A dictionary containing results for each target group with 3 keys.
+
```
Comment (xianyuanliu) [inline kale/evaluate/cross_validation.py:24]: Is this blank line necessary?

File: kale/evaluate/cross_validation.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    enc = OneHotEncoder(handle_unknown="ignore")
+    group_mat = enc.fit_transform(groups.reshape(-1, 1)).toarray()
+    target, num_samples, accuracy = np.unique(groups).tolist(), [], []
+
+    logo = LeaveOneGroupOut()
+    for train, test in logo.split(x, y, groups=groups):
+        x_src, x_tgt = x[train], x[test]
+        y_src, y_tgt = y[train], y[test]
+
+        if domain_adaptation:
```
[the remark this answers is not in the document]
Comment (xianyuanliu) [reply kale/evaluate/cross_validation.py:37]: Should we use the term `is_domain_adaptation` or `is_da` referring [here](https://github.com/pykale/pykale/blob/main/kale/pipeline/domain_adapter.py#L98)? @shuo-zhou

**Review by xianyuanliu — CHANGES_REQUESTED** [review]
Need to be changed or discussed

File: examples/multisite_neuroimg_adapt/main.py (hunk 11 above)
Comment (shuo-zhou) [reply examples/multisite_neuroimg_adapt/main.py:54]: `SITE_ID` is a column name of the phenotype data

File: kale/evaluate/cross_validation.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (xianyuanliu) [reply kale/evaluate/cross_validation.py:36]: `use_domain_adaptation` is better.

File: examples/multisite_neuroimg_adapt/main.py [hunk 14]
```diff
@@ -2,13 +2,13 @@
 Autism Detection: Domain Adaptation for Multi-Site Neuroimaging Data Analysis
 
 Reference:
-[1] Cameron Craddock, Yassine Benhajali, Carlton Chu, Francois Chouinard, Alan Evans, András Jakab, Budhachandra Singh Khundrakpam, John David Lewis, Qingyang Li, Michael Milham, Chaogan Yan, Pierre Bellec (2013). The Neuro Bureau Preprocessing Initiative: open sharing of preprocessed neuroimaging data and derivatives. In Neuroinformatics 2013, Stockholm, Sweden.
+[1] Cameron, C., Yassine, B., Carlton, C., Francois, C., Alan, E., András, J., Budhachandra, K., John, L., Qingyang, L., Michael, M., Chaogan, Y. and Pierre, B. (2013). The Neuro Bureau Preprocessing Initiative: open sharing of preprocessed neuroimaging data and derivatives. Frontiers in Neuroinformatics, 7. https://doi.org/10.3389/conf.fninf.2013.09.00041
```
Comment (xianyuanliu) [inline examples/multisite_neuroimg_adapt/main.py:5]: Mind the capitalization of the four titles.

File: examples/multisite_neuroimg_adapt/README.md [hunk 15]
```diff
@@ -33,14 +33,10 @@ autism detection.
 
 ### References
 
-[1] Cameron Craddock, Yassine Benhajali, Carlton Chu, Francois Chouinard, Alan Evans, András Jakab, Budhachandra Singh
-Khundrakpam, John David Lewis, Qingyang Li, Michael Milham, Chaogan Yan, Pierre Bellec (2013). The Neuro Bureau
-Preprocessing Initiative: open sharing of preprocessed neuroimaging data and derivatives. In *Neuroinformatics 2013*,
-Stockholm, Sweden.
+[1] Cameron, C., Yassine, B., Carlton, C., Francois, C., Alan, E., András, J., Budhachandra, K., John, L., Qingyang, L., Michael, M., Chaogan, Y. and Pierre, B. (2013). [The Neuro Bureau Preprocessing Initiative: open sharing of preprocessed neuroimaging data and derivatives](https://doi.org/10.3389/conf.fninf.2013.09.00041). Frontiers in Neuroinformatics, 7.
```
Comment (xianyuanliu) [inline examples/multisite_neuroimg_adapt/README.md:36]: Mind the capitalization of the four titles.

File: examples/multisite_neuroimg_adapt/tutorial.ipynb [hunk 16]
```diff
@@ -25,13 +25,13 @@
         "\n",
         "**Reference:**\n",
         "\n",
-        "[1] Cameron Craddock, Yassine Benhajali, Carlton Chu, Francois Chouinard, Alan Evans, Andr\u00e1s Jakab, Budhachandra Singh Khundrakpam, John David Lewis, Qingyang Li, Michael Milham, Chaogan Yan, Pierre Bellec (2013). The Neuro Bureau Preprocessing Initiative: open sharing of preprocessed neuroimaging data and derivatives. In *Neuroinformatics 2013*, Stockholm, Sweden.\n",
+        "[1] Cameron, C., Yassine, B., Carlton, C., Francois, C., Alan, E., Andr\u00e1s, J., Budhachandra, K., John, L., Qingyang, L., Michael, M., Chaogan, Y. and Pierre, B. (2013). [The Neuro Bureau Preprocessing Initiative: open sharing of preprocessed neuroimaging data and derivatives](https://doi.org/10.3389/conf.fninf.2013.09.00041). Frontiers in Neuroinformatics, 7.\n",
```
Comment (xianyuanliu) [inline examples/multisite_neuroimg_adapt/tutorial.ipynb:28]: The same. Capitalization.

**Review by xianyuanliu — CHANGES_REQUESTED** [review]
Well done. Minor changes are needed.

**Review by xianyuanliu — APPROVED** [review]
Well done. LGTM.

**Review by haipinglu — APPROVED** [review]
Well done!
